### PR TITLE
Standalone extension codestarts

### PR DESCRIPTION
**Pull request by ia3andy**
- Updated docs
- example repo: https://github.com/ia3andy/aloha-code
- Add `CodestartResourceLoadersBuilder` to replace the different factories which became hard to use

### CODE REVIEW

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/CodestartResourceLoadersBuilder.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    private CodestartResourceLoadersBuilder() {
+    }
+
+    public CodestartResourceLoadersBuilder catalog(ExtensionCatalog catalog) {
+        this.catalog = catalog;
+        return this;
+    }
+
+    public CodestartResourceLoadersBuilder mvnArtifactResolver(MavenArtifactResolver mvnArtifactResolver) {
```
Comment (aloubyansky) [inline independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/CodestartResourceLoadersBuilder.java:27]: Simply `artifactResolver` would be better.

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/CodestartResourceLoadersBuilder.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import io.quarkus.registry.catalog.ExtensionCatalog;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public final class CodestartResourceLoadersBuilder {
+    ExtensionCatalog catalog = null;
+    MavenArtifactResolver mvnArtifactResolver = QuarkusProjectHelper.artifactResolver();
+    String baseCodestartsArtifactCoords = BASE_CODESTARTS_ARTIFACT_COORDS;
+    Collection<String> extraCodestartsArtifactCoords = new ArrayList<>();
```
Comment (aloubyansky) [inline independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/CodestartResourceLoadersBuilder.java:17]: is there a reason these aren't private?

File: independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java [hunk 3]
```diff
@@ -401,23 +392,18 @@ private void completeCodestartArtifact(ObjectMapper mapper, ObjectNode extObject
...
+        codestartObject.put("artifact", codestartArtifactCoords.toString());
         if (!skipCodestartValidation) {
             // first we look for it in the workspace, if it's in there we don't need to actually resolve the artifact, because it might not have been built yet
-            if (workspaceProvider.workspace().getProject(codestartArtifact.getGroupId(),
-                    codestartArtifact.getArtifactId()) == null) {
+            if (workspaceProvider.workspace().getProject(codestartArtifactCoords.getGroupId(),
+                    codestartArtifactCoords.getArtifactId()) == null) {
                 try {
-                    resolve(codestartArtifact);
+                    resolve(new DefaultArtifact(codestartArtifactCoords.toString()));
```
Comment (ia3andy) [inline independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java:404]: @aloubyansky maybe better to use `codestartArtifact` here no?

Comment (ia3andy) [reply independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java]: Done

File: independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java [hunk 4]
```diff
@@ -401,23 +392,21 @@ private void completeCodestartArtifact(ObjectMapper mapper, ObjectNode extObject
...
             return;
         }
-        org.eclipse.aether.artifact.Artifact codestartArtifact = DependencyNodeUtils.toArtifact(mvalue.asText());
 
-        if (codestartArtifact.getVersion() == null || codestartArtifact.getVersion().isEmpty()) {
-            codestartArtifact = codestartArtifact.setVersion(project.getVersion());
-            codestartObject.put("artifact",
-                    codestartArtifact.getGroupId() + ":" + codestartArtifact.getArtifactId() + ":"
-                            + codestartArtifact.getClassifier() + ":" + codestartArtifact.getExtension() + ":"
-                            + codestartArtifact.getVersion());
```
Comment (aloubyansky) [inline independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java:411]: Do you mind adding a comment about what is expected here and what is not supported?

Comment (ia3andy) [reply independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java:411]: @aloubyansky sure, in another PR?

"G:A" or anything supported by AppArtifactCoords.fromString() with `${project.version}` support.

Comment (aloubyansky) [reply independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java:411]: Doesn't look like it's worth creating a separate PR for a comment. Could you add it to this one?

Comment (ia3andy) [reply independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java:411]: Ok everything's fine beside that?

Comment (ia3andy) [reply independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java:411]: I just hate to wait for the CI for a comment

Comment (aloubyansky) [reply independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java:411]: Yes, looks good, thanks!

**Review by aloubyansky — APPROVED** [review]

**Pull request merged**